netket/variational.py

```diff
@@ -1,31 +1,131 @@
+        self,
+        hamiltonian,
+        sampler,
+        optimizer,
+        n_samples,
+        discarded_samples=None,
+        discarded_samples_on_init=0,
+        target="energy",
+        method="Sr",
+        diag_shift=0.01,
+        use_iterative=False,
+        use_cholesky=None,
+        sr_lsq_solver="LLT",
+    ):
+
+        self._mynode = _MPI.rank()
+        self.machine = sampler.machine
+
+        if method == "Gd":
+            self.sr = None
+            self._vmc = _Vmc(
+                hamiltonian=hamiltonian,
+                sampler=sampler,
+                optimizer=optimizer,
+                n_samples=n_samples,
+                n_discard=discarded_samples,
+                sr=None,
+            )
+        elif method == "Sr":
+            self.sr = _SR(
+                lsq_solver=sr_lsq_solver,
+                diag_shift=diag_shift,
+                use_iterative=use_iterative,
+                is_holomorphic=sampler.machine.is_holomorphic,
+            )
+            self._vmc = _Vmc(
+                hamiltonian=hamiltonian,
+                sampler=sampler,
+                optimizer=optimizer,
+                n_samples=n_samples,
+                n_discard=discarded_samples,
+                sr=self.sr,
+            )
+        else:
+            raise ValueError("Allowed method options are Gd and Sr")
+
+        if use_cholesky and sr_lsq_solver != "LLT":
+            raise ValueError(
+                "Inconsistent options specified: `use_cholesky && sr_lsq_solver != 'LLT'`."
+            )
+
+        if discarded_samples_on_init != 0:
+            warnings.warn(
+                "discarded_samples_on_init does not have any effect and should not be used",
+                DeprecationWarning,
+            )
+
+        self.advance = self._vmc.advance
+        self.add_observable = self._vmc.add_observable
+        self.get_observable_stats = self._vmc.get_observable_stats
+        self.reset = self._vmc.reset
+
+        warnings.warn(
+            "netket.variational.Vmc will be deprecated in version 3, use netket.Vmc instead",
+            PendingDeprecationWarning,
+        )
+
+    def _add_to_json_log(self, step_count):
+
+        stats = self.get_observable_stats()
+        self._json_out["Output"].append({})
+        self._json_out["Output"][-1] = {}
+        json_iter = self._json_out["Output"][-1]
+        json_iter["Iteration"] = step_count
+        for key, value in stats.items():
+            st = value.asdict()
+            st["Mean"] = st["Mean"].real
+            json_iter[key] = st
+
+    def _init_json_log(self):
+
+        self._json_out = {}
+        self._json_out["Output"] = []
+
+    def run(
+        self, output_prefix, n_iter, step_size=1, save_params_every=50, write_every=50
+    ):
+        self._init_json_log()
+
+        for k in range(n_iter):
+            self.advance(step_size)
+
+            self._add_to_json_log(k)
+            if k % write_every == 0 or k == n_iter - 1:
```

Comment:
In the C++ `Vmc` class, `vmc.run(N, save_params_every=N)` has the counter-intuitive result of only storing the first step `i=0`, whereas `vmc.run(N+1, save_params_every=N)` would have been correct. Fixing this, as this class does by checking `k == n_iter - 1`, is indeed a good thing.

Reply:
Indeed that was an issue!